src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java

```diff
@@ -0,0 +1,178 @@
+     * and read mapping quality.  These per read items are delimited with an '@' character.
+     */
+    @Argument(fullName = "showVerbose", shortName = "verbose", doc = "Add an extra verbose section to the pileup output", optional = true)
+    public boolean showVerbose = false;
+
+    /**
+     * This enables annotating the pileup to show overlaps with metadata from a Feature file(s). For example, if you provide a
+     * VCF and there is a SNP at a given location covered by the pileup, the pileup output at that position will be
+     * annotated with the corresponding source Feature identifier.
+     */
+    @Argument(fullName = "metadata", shortName = "metadata", doc = "Features file(s) containing metadata", optional = true)
+    public List<FeatureInput<Feature>> metadata = new ArrayList<>();
+
+    /**
+     * Adds the length of the insert each base comes from to the output pileup. Here, "insert" refers to the DNA insert
+     * produced during library generation before sequencing.
+     */
+    @Hidden
+    @Argument(fullName = "outputInsertLength", shortName = "outputInsertLength", doc = "Output insert length", optional = true)
+    public boolean outputInsertLength = false;
+
+    @Override
+    public void apply(AlignmentContext alignmentContext, ReferenceContext referenceContext, FeatureContext featureContext) {
+        final String features = getFeaturesString(featureContext);
+        final ReadPileup basePileup = alignmentContext.getBasePileup();
+        final StringBuilder s = new StringBuilder();
+        s.append(String.format("%s %s",
+                basePileup.getPileupString((hasReference()) ? (char) referenceContext.getBase() : 'N'),
+                features));
+        if (outputInsertLength) {
+            s.append(" ").append(insertLengthOutput(basePileup));
+        }
+        if (showVerbose) {
+            s.append(" ").append(createVerboseOutput(basePileup));
+        }
+        s.append("\n");
+        out.print(s.toString());
+    }
+
+    /**
+     * Get a string representation for the metadata
+     *
+     * @param featureContext Context for the features.
+     *
+     * @return String representation of the metadata
+     */
+    private String getFeaturesString(final FeatureContext featureContext) {
+        List<String> featuresStrings = new ArrayList<>();
+        for (final Feature datum : featureContext.getValues(metadata)) {
+            featuresStrings.add(datum.toString());
+        }
+        String featuresString = Utils.join(", ", featuresStrings);
+        if (!featuresString.equals("")) {
+            featuresString = "[Feature(s): " + featuresString + "]";
+        }
+        return featuresString;
+    }
+
+    /**
+     * Format the insert length for a pileup
+     * @param pileup the pileup to format
+     * @return a comma-separated string with insert lengths
+     */
+    private static String insertLengthOutput(final ReadPileup pileup) {
```

Comment:
change into a package-visible method, add `@VisibleForTesting` and add unit tests for it